File: Lab2/ex1.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-

from mpl_toolkits.mplot3d import Axes3D
import matplotlib.pyplot as plt
from matplotlib import cm
import scipy.io as spio
from scipy.stats import multivariate_normal
import numpy as np
import pandas as pd

## DATA PREPARATION

mat = spio.loadmat('heightWeight.mat',squeeze_me=True)
raw_data = mat['heightWeightData']
m_X = raw_data[raw_data[:,0]==1,:]
f_X = raw_data[raw_data[:,0]==2,:] 

plt.figure(1)
plt.scatter(m_X[:,1],m_X[:,2],marker='x',c='b') # Males 
plt.scatter(f_X[:,1],f_X[:,2],marker='o',c='r') # Females
plt.title('Males and Females Data')
plt.xlabel('Height (cm)')
plt.ylabel('Weight (kg)')

plt.figure(2)
plt.scatter(m_X[:,1],m_X[:,2],marker='x',c='b') # Males 
plt.title('Males Data')
plt.xlabel('Height (cm)')
plt.ylabel('Weight (kg)')

plt.figure(3)
plt.scatter(f_X[:,1],f_X[:,2],marker='o',c='r') # Females
plt.title('Females Data')
plt.xlabel('Height (cm)')
plt.ylabel('Weight (kg)')

plt.show()

""" Histograms """
plt.figure(4)
plt.hist(m_X[:,2],11,alpha=0.5,color='b')
plt.hist(f_X[:,2],11,alpha=0.5,color='r')
plt.title('Males & Females Weight Histogram')
plt.xlabel('weight')
plt.ylabel('frequency')
plt.grid(True)
plt.show()

plt.figure(5)
plt.hist(m_X[:,1],11,alpha=0.5,color='b')
plt.hist(f_X[:,1],11,alpha=0.5,color='r')
plt.title('Males & Females Height Histogram')
plt.xlabel('Height')
plt.ylabel('frequency')
plt.grid(True)
plt.show()

""" P(X|y=c,u,E) Maximum Likelihood of mean and Cov"""
#x_m = m_X[:,1:3]

m_mean = np.mean(m_X[:,1:3],axis=0)
f_mean = np.mean(f_X[:,1:3],axis=0)
m_cov = np.cov(m_X[:,1:3],rowvar=False)
f_cov = np.cov(f_X[:,1:3],rowvar=False)
""" MALES """
x1 = range(120,220,1)
x2 = range(30,140,1)
XM1,XM2 = np.meshgrid(x1,x2)
B = np.array([XM1.flatten(),XM2.flatten()])
M = multivariate_normal.pdf(B.T, mean=m_mean, cov=m_cov)
M = M.reshape(len(x2),len(x1))
fig = plt.figure(6)
ax = fig.add_subplot(111, projection='3d')
# Plot the surface
ax.plot_surface(XM1, XM2, M, color='b',rstride=1, cstride=1, cmap=cm.jet,
linewidth=0, antialiased=False)
plt.title('Males Joing PDF')
plt.show()


""" FEMALES """

x1 = range(120,220,1)
x2 = range(30,140,1)
XF1,XF2 = np.meshgrid(x1,x2)
B = np.array([XF1.flatten(),XF2.flatten()])
F = multivariate_normal.pdf(B.T, mean=f_mean, cov=f_cov)


F = F.reshape(len(x2),len(x1))
fig = plt.figure(6)
ax = fig.add_subplot(111, projection='3d')
# Plot the surface
ax.plot_surface(XF1, XF2, F, color='b',rstride=1, cstride=1, cmap=cm.jet,
linewidth=0, antialiased=False)
plt.title('Females Joing PDF')
#ax.set_zlim3d(-1.01, 1.01);
plt.show()
plt.figure(9)
plt.figure(1)
plt.scatter(m_X[:,1],m_X[:,2],marker='x',c='b') # Males 
plt.scatter(f_X[:,1],f_X[:,2],marker='o',c='r') # Females
plt.contour(XM1, XM2, M, linewidths=1, colors='k')
plt.contour(XF1, XF2, F, linewidths=1, colors='k')
plt.title('Joint Multivariant Normal PDF')
plt.xlabel('Height (cm)')
plt.ylabel('Weight (kg)')
plt.show()
#m_cov_MLE = np.sum(x_m.dot(x_m.T))*(1/len(x_m))
#m_cov_MLE -= m_mean.dot(m_mean.T)

#X_train = np.asarray(mat[''])
#y_train = np.asarray(mat['ytrain'])

File: Lab2/ex1.m
% LAB-2 Ex-1
% DATA Import
rawData=importdata('heightWeight.mat'); %call for import the file name
males = rawData(:,1)==1;

maleDataSet=rawData(males,:);
malesRaw = maleDataSet;
females = rawData(:,1)==2;
femaleDataSet=rawData(females,:);
malesTotal=size(maleDataSet,1);
femalesTotal=size(femaleDataSet,1);
% ---------------- % 

figure
subplot(2,3,1) 
% male Data in scatter plot
scatter(maleDataSet(:,2),maleDataSet(:,3));
xlabel('Height');
ylabel('Weight');
title('Males: H vs W');

%plot Height of Males as historgam
subplot(2,3,2);
hist(maleDataSet(:,2))
xlabel('Height');
ylabel('Occurences');
title('Males: Weight');
%plot Weight of Males as histogram
subplot(2,3,3);
hist(maleDataSet(:,3))
xlabel('Weight');
ylabel('Occurences');
title('Males: Weight');

%Female Data in scatter plot
subplot(2,3,4)
scatter(femaleDataSet(:,2),femaleDataSet(:,3));
xlabel('Height');
ylabel('Weight');
title('Females: H vs W');

%plot Height of Females as historgam
subplot(2,3,5);
hist(femaleDataSet(:,2))
xlabel('Height');
ylabel('Occurences');
title('Females: Height');
%plot Weight of Females as histogram
subplot(2,3,6);
hist(femaleDataSet(:,3))
xlabel('Weight');
ylabel('Occurences');
title('Females: Weight');

% Calculate the mean Vector of Males data set
% Vector contains the mean of Height and Weight
meanMale=mean(maleDataSet(:,2:3));
% Calculate the mean Vector of Females data set
% Vector contains the mean of Height and Weight
meanFemale=mean(femaleDataSet(:,2:3));

% maleDataSet(:,2:3) = maleDataSet(:,2:3) - meanMale;
% varMales = var(malesRaw(:,2:3));
% maleDataSet(:,2:3)=maleDataSet(:,2:3)./sqrt(varMales);

runningMatrix = zeros(2,2); % Define a 2x2 empty matrix 
for i=1:malesTotal
    runningMatrix = runningMatrix + maleDataSet(i,2:3)'*maleDataSet(i,2:3);
end

% Calculation of the Covariant Matrix of Males 
covMatrix_males=(1/malesTotal)*runningMatrix - meanMale'*meanMale;

runningMatrix = zeros(2,2);% Define a 2x2 empty matrix 
for i=1:femalesTotal
    runningMatrix = runningMatrix + femaleDataSet(i,2:3)'*femaleDataSet(i,2:3);
end

% Calculation of covarient Matrix of Females
covMatrix_females=(1/femalesTotal)*runningMatrix - meanFemale'*meanFemale;

% Plot the 2-D joint pdf for Males 
figure
x1 = 160:1:205; x2 = 50:1:130;
[X1,X2] = meshgrid(x1,x2);
F = mvnpdf([X1(:) X2(:)],meanMale,covMatrix_males);
F = reshape(F,length(x2),length(x1));
surf(x1,x2,F);
caxis([min(F(:))-.5*range(F(:)),max(F(:))]);
% axis([160 205 50 230 0 max(F(:))])
xlabel('weight'); ylabel('height'); zlabel('Probability Density');
title('Males: Joint PDF');

% Plot the 2-D joint pdf for Females  

figure
x1 = 160:1:205; x2 = 50:1:130;
[X1,X2] = meshgrid(x1,x2);
F = mvnpdf([X1(:) X2(:)],meanFemale,covMatrix_females);
F = reshape(F,length(x2),length(x1));
surf(x1,x2,F);
caxis([min(F(:))-.5*range(F(:)),max(F(:))]);
% axis([160 205 50 230 0 max(F(:))])
xlabel('weight'); ylabel('height'); zlabel('Probability Density');
title('Females: Joint PDF');